# Project Fine-tune and Evaluate LLM model on popular benchmark dataset

- I want to evaluate different LLM on benchmark datasets to identify the strengths and weaknesses. The benchmark datasets can be used from the research I did with Mishra. Then, I can implement open-source LLM on these benchmark datasets.
- I think I will learn a lot on the evaluation aspect of how to determine if the newly created model is good.
- Also, I think I can do a simple model in which I fine tune the baseline models to improve the accuracy and precision.


### Step 1: Choosing BenchMark dataset and download from hugging face

- In this project, I use GLUE dataset (MRPC) to benchmark the model.

- MRPC consists of sentences taken from new sources. Each pair of sentences is labeled as either a paraphrase or not. The purpose is to identify the pair of sentences expressing the same meaning or not.

- In examining the dataset, the columns sentence 1, sentence 2 and label indicate whether sentence 1 and sentence 2 are paraphrased of each other or not.

In [ ]:
# Choose benchmark dataset GLUE, SQAD
!pip install transformers
!pip install torch

  Using cached nvidia_cuda_nvrtc_cu12-12.1.105-py3-none-manylinux1_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_cuda_runtime_cu12-12.1.105-py3-none-manylinux1_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_cuda_cupti_cu12-12.1.105-py3-none-manylinux1_x86_64.whl.metadata (1.6 kB)
  Using cached nvidia_cudnn_cu12-8.9.2.26-py3-none-manylinux1_x86_64.whl.metadata (1.6 kB)
  Using cached nvidia_cublas_cu12-12.1.3.1-py3-none-manylinux1_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_cufft_cu12-11.0.2.54-py3-none-manylinux1_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_curand_cu12-10.3.2.106-py3-none-manylinux1_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_cusolver_cu12-11.4.5.107-py3-none-manylinux1_x86_64.whl.metadata (1.6 kB)
  Using cached nvidia_cusparse_cu12-12.1.0.106-py3-none-manylinux1_x86_64.whl.metadata (1.6 kB)
  Using cached nvidia_nccl_cu12-2.20.5-py3-none-manylinux2014_x86_64.whl.metadata (1.8 kB)
  Using cached nvidia_nvtx_cu12-12.1.105-py3-none-manylinu

In [ ]:
!pip install datasets

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 547.8/547.8 kB 34.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 11.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 316.1/316.1 kB 29.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 39.9/39.9 MB 55.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 64.9/64.9 kB 6.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 14.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.1/194.1 kB 19.7 MB/s eta 0:00:00
  Attempting uninstall: requests
    Found existing installation: requests 2.31.0
    Uninstalling requests-2.31.0:
      Successfully uninstalled requests-2.31.0
  Attempting uninstall: pyarrow
    Found existing installation: pyarrow 14.0.2
    Uninstalling pyarrow-14.0.2:
      Successfully uninstalled pyarrow-14.0.2
  Attempting uninstall: fsspec
    Found existing installation: fsspec 2024.6.1
    Uninstalling fsspec-2024.6.1:
  

In [ ]:
from datasets import load_dataset

dataset = load_dataset('glue', 'mrpc')
dataset


/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


Generating train split:   0%|          | 0/3668 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/408 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/1725 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['sentence1', 'sentence2', 'label', 'idx'],
        num_rows: 3668
    })
    validation: Dataset({
        features: ['sentence1', 'sentence2', 'label', 'idx'],
        num_rows: 408
    })
    test: Dataset({
        features: ['sentence1', 'sentence2', 'label', 'idx'],
        num_rows: 1725
    })
})

### Step 2: Choosing Open Source Baseline Model

In [ ]:
from transformers import BertTokenizer, BertForSequenceClassification
import torch

# Load pre-trained BERT model and tokenizer
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
model = BertForSequenceClassification.from_pretrained('bert-base-uncased')



tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [ ]:
dataset

DatasetDict({
    train: Dataset({
        features: ['sentence1', 'sentence2', 'label', 'idx'],
        num_rows: 3668
    })
    validation: Dataset({
        features: ['sentence1', 'sentence2', 'label', 'idx'],
        num_rows: 408
    })
    test: Dataset({
        features: ['sentence1', 'sentence2', 'label', 'idx'],
        num_rows: 1725
    })
})

In [ ]:
# Preprocess the data

def tokenize_function(examples):
    return tokenizer(examples['sentence1'], examples['sentence2'], padding='max_length', truncation=True)

tokenized_datasets = dataset.map(tokenize_function, batched = True)

### Step 3: Evaluate the baseline model on the benchmark dataset to set a reference point

In [ ]:
from transformers import Trainer

# Define a trainer with no training arguments (only evaluation)
trainer = Trainer(
    model=model,
    eval_dataset=tokenized_datasets['validation']
)

In [ ]:
results = trainer.evaluate()
print(results)

{'eval_loss': 0.6237205862998962, 'eval_runtime': 5.0478, 'eval_samples_per_second': 80.828, 'eval_steps_per_second': 10.104}


### Step 4: Fine tune the baseline model


- evaluation_strategy: allows you to monitor the performance of the model.
- learning rate = 2e-5: the step size fro the optimizer to update the model weights. A smaller learning rates help achieving convergence better by making small controlled updates to the model weight
- num_train_epochs: specify the number of the times the model go through the entire dataset
- per_device_train_batch_size: set the batch size for training
- weights_step = 500: number of steps for the learning rate to warm up before reaching the intial learning rate
- weight_decay = 0.01: add penalty to the loss
- logging_dir = './logs': specify the log where they are stored
- logging_steps = 10: number of steps between the logging updates.

What we fine-tuned:
- Optimized learning rate: adjusting the learning and using warmup steps can prevent issues like overshooting and slow convergence. (adjusting the learning rate to a suitable number)
- Regularization and monitoring: add penalty loss to reduce overfitting.
- Batch size and epochs: proper setting of batch size and number of epochs ensures a balance between training efficiency and model performance. Larger batch sizes and epochs requires a lot of computational resources.

In [ ]:
!pip install transformers

In [ ]:
from transformers import TrainingArguments, Trainer

# Specify output_dir, leraning_rate, evaluation_strategy, num_train_epoch, weight decay
training_args = TrainingArguments(
    output_dir='./results',
    evaluation_strategy='epoch',
    learning_rate=2e-5,
    num_train_epochs=3,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    warmup_steps=500,
    weight_decay=0.01,
    logging_dir='./logs',
    logging_steps=10,
)

fine_tune_trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_datasets['train'],
    eval_dataset=tokenized_datasets['validation'],
)

fine_tune_trainer.train()

/usr/local/lib/python3.10/dist-packages/transformers/training_args.py:1494: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(


TrainOutput(global_step=690, training_loss=0.43431156493615414, metrics={'train_runtime': 238.1998, 'train_samples_per_second': 46.197, 'train_steps_per_second': 2.897, 'total_flos': 2895274053181440.0, 'train_loss': 0.43431156493615414, 'epoch': 3.0})

### Step 5: Evaluate the baseline model on the benchmark dataset to determine if that is better.
A step before this can involve understanding the fine-tuning process and how to make the model become better. Research about different methods that we can achieve that

In [ ]:
fine_tune_results = fine_tune_trainer.evaluate()
print(fine_tune_results)

{'eval_loss': 0.39543601870536804, 'eval_runtime': 2.8832, 'eval_samples_per_second': 141.509, 'eval_steps_per_second': 9.018, 'epoch': 3.0}


### Step 6: Save the model

In [ ]:
model.save_pretrained('./fine_tuned_model')
tokenizer.save_pretrained('./fine_tuned_model')


fine_tune_trainer.save_model('./fine_tuned_model')


In [ ]:
# give access to google drive
from google.colab import drive
drive.mount('/content/drive')

# path
outputPath = '/content/drive/MyDrive/Summer 2024/projects/project1'

model.save_pretrained(outputPath)
tokenizer.save_pretrained(outputPath)
fine_tune_trainer.save_model(outputPath)


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


### Step 7: Report and Summaries


- Without fine-tuning:
{'eval_loss': 0.6237205862998962, 'eval_runtime': 5.0478, 'eval_samples_per_second': 80.828, 'eval_steps_per_second': 10.104}


- With finetuning:
{'eval_loss': 0.39543601870536804, 'eval_runtime': 2.8832, 'eval_samples_per_second': 141.509, 'eval_steps_per_second': 9.018, 'epoch': 3.0}


- Without Fine-Tuning:
Evaluation Loss: 0.6237
Evaluation Runtime: 5.0478 seconds
Samples per Second: 80.828
Steps per Second: 10.104

- With Fine-tuning:
Evaluation Loss: 0.3954 (improved from 0.6237)
Evaluation Runtime: 2.8832 seconds (faster than 5.0478 seconds)
Samples per Second: 141.509 (faster than 80.828 samples per second)
Steps per Second: 9.018 (slightly slower than 10.104 steps per second)
Epochs: 3.0 (indicates the number of training epochs)

- Result:
- The fine-tuned model is approximately 42.9% faster in evaluation runtime compared to the baseline model
- The fine-tuned model shows a 36.6% improvement in evaluation loss
- Pushed to hugging face hub for open-source collaboration


What's next:
- Add another model: T-5
- Add another benchmark dataset and evaluate better with the benchmark dataset



In [ ]:
!pip install transformers huggingface_hub


In [ ]:
# Save to hugging face
# Access token:

from huggingface_hub import login

login()



In [ ]:
fine_tune_trainer.push_to_hub('fine_tuned_bert')

Upload 2 LFS files:   0%|          | 0/2 [00:00<?, ?it/s]

model.safetensors:   0%|          | 0.00/438M [00:00<?, ?B/s]

training_args.bin:   0%|          | 0.00/5.05k [00:00<?, ?B/s]

CommitInfo(commit_url='https://huggingface.co/Thang203/results/commit/b31469c1654c5e5b097ca49fe5ee682374009935', commit_message='fine_tuned_bert', commit_description='', oid='b31469c1654c5e5b097ca49fe5ee682374009935', pr_url=None, pr_revision=None, pr_num=None)